llvm/lib/Target/AIE/AIE2AddrSpace.h

```diff
@@ -0,0 +1,46 @@
+//===-- AIE2AddrSpace.h - Define Address Space for AIEngine V2 ---*- C++-*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares the AIEngine V2 Address Space and DM banks
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SUPPORT_AIE2ADDRSPACE_H
+#define LLVM_SUPPORT_AIE2ADDRSPACE_H
+
+namespace llvm {
+
+namespace AIE2 {
+
+enum class AddressSpaces {
+  none, // Default address space
+  PM,   // Address space for Program Memory (PM)
+  DM,   // Address space for Data Memory(DM) includes Bank A, B, C, D
+  DM_test,
+  stack, // Address space for stack
+  a,     // Address space for DM Bank A
+  b,     // Address space for DM Bank B
+  c,     // Address space for DM Bank C
+  d,     // Address space for DM Bank D
+  ab,    // Address space for DM Bank A and B
+  ac,    // Address space for DM Bank A and C
+  ad,    // Address space for DM Bank A and D
+  bc,    // Address space for DM Bank B and C
+  bd,    // Address space for DM Bank B and D
+  cd,    // Address space for DM Bank C and D
+  TM     // Address space for TM (Tile Memory)
+};
```

Comment:
I would like to have a clarifying comment. AFAICS, we assume conflict between e.g. AB and BD, which I guess is safe. I just don't know what the programmer may assume and what the compiler may try to optimize when declaring an object in e.g. BD.
I could imagine that using analysis the compiler could try to specialize a BD object to D in order not to make it conflict with AB. It would then need to retype all derived pointers to D. 
Alternatively, a BD object somehow straddles both B and D, which means it cannot be specialize to D. 
'for DM Bank B and D' sounds ambiguous to me.

Reply:
Based on my understanding of how buffers are allocated in Data Memory of AIE Tile, they are in Ping-Pong fashion where in Ping is in a different bank Pong is in different bank. That is where AB, BD , etc comes in play.  In case say the IFM is on bank 0 & bank 1 as per mapper constrain the annotation would be AB. 

At least from the IFM, WFM, OFM buffer we cannot optimize the buffer using the compiler to a single bank. While one bank say Ping is being read by core , usually Pong is either read out/in by the DMA.